Add missing type hints to `dbt/core/version.py`

### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

We're putting some effort into improving our typing. There are two type errors in `dbt/core/version.py`, we should fix them.

```
core/dbt/version.py:52: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
core/dbt/version.py:211: error: Unused "type: ignore" comment  [unused-ignore]
```

### Acceptance criteria

`dbt/core/verison.py` doesn't have any type errros

### Suggested Tests

N/A

### Impact to Other Teams

None

### Will backports be required?

No

### Context

_No response_

--- a/.changes/unreleased/Under the Hood-20240827-105014.yaml
+++ b/.changes/unreleased/Under the Hood-20240827-105014.yaml
@@ -1,0 +1,6 @@
+kind: Under the Hood
+body: Additional type hints for `core/dbt/version.py`
+time: 2024-08-27T10:50:14.047859-05:00
+custom:
+  Author: QMalcolm
+  Issue: "10612"

--- a/core/dbt/version.py
+++ b/core/dbt/version.py
@@ -49,7 +49,10 @@ def get_latest_version(
     return semver.VersionSpecifier.from_version_string(version_string)
 
 
-def _get_core_msg_lines(installed, latest) -> Tuple[List[List[str]], str]:
+def _get_core_msg_lines(
+    installed: semver.VersionSpecifier,
+    latest: Optional[semver.VersionSpecifier],
+) -> Tuple[List[List[str]], str]:
     installed_s = installed.to_version_string(skip_matcher=True)
     installed_line = ["installed", installed_s, ""]
     update_info = ""
@@ -208,7 +211,7 @@ def _get_dbt_plugins_info() -> Iterator[Tuple[str, str]]:
         except ImportError:
             # not an adapter
             continue
-        yield plugin_name, mod.version  # type: ignore
+        yield plugin_name, mod.version
 
 
 def _get_adapter_plugin_names() -> Iterator[str]:

